### PR TITLE
Thread safe fix

### PR DESCRIPTION
**Pull request by lebron374**
## What is the purpose of the change

IndexService#getAndCreateLastIndexFile threadsafe issue

## Brief changelog

XX

## Verifying this change

XXXX

Follow this checklist to help us incorporate your contribution quickly and easily. Notice, `it would be helpful if you could finish the following 5 checklist(the last one is not necessary)before request the community to review your PR`.

- [x] Make sure there is a [Github issue](https://github.com/apache/rocketmq/issues/1986) filed for the change (usually before you start working on it). Trivial changes like typos do not require a Github issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue. 
- [x] Format the pull request title like `[ISSUE #123] Fix UnknownException when host config not exist`. Each commit in the pull request should have a meaningful subject line and body.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit-test(over 80% coverage) to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/apache/rocketmq/tree/master/test).
- [x] Run `mvn -B clean apache-rat:check findbugs:findbugs checkstyle:checkstyle` to make sure basic checks pass. Run `mvn clean install -DskipITs` to make sure unit-test pass. Run `mvn clean test-compile failsafe:integration-test`  to make sure integration-test pass.
- [ ] If this contribution is large, please file an [Apache Individual Contributor License Agreement](http://www.apache.org/licenses/#clas).

### CODE REVIEW

File: store/src/main/java/org/apache/rocketmq/store/index/IndexService.java [hunk 1]
```diff
@@ -313,14 +313,24 @@ public IndexFile getAndCreateLastIndexFile() {
...
-                        lastUpdateIndexTimestamp);
                 this.readWriteLock.writeLock().lock();
-                this.indexFileList.add(indexFile);
+
+                IndexFile tmp = null;
+                if (!this.indexFileList.isEmpty()) {
+                    tmp = this.indexFileList.get(this.indexFileList.size() - 1);
+                }
+
+                if (null == tmp || tmp.isWriteFull()) {
```
Comment (imaffe) [inline store/src/main/java/org/apache/rocketmq/store/index/IndexService.java:323]: should we update `prevIndexFile` when `tmp.isWriteFull()` is true ?

File: store/src/test/java/org/apache/rocketmq/store/StoreStatsServiceTest.java [hunk 2]
```diff
@@ -22,9 +22,18 @@
...
 import org.junit.Test;
+import static org.junit.Assert.*;
 
 public class StoreStatsServiceTest {
 
+    @Test
+    public void StoreStatsServiceStringFromat() throws Exception {
+
+        StoreStatsService storeStatsService = new StoreStatsService();
+        new Thread(storeStatsService).start();
```
Comment (imaffe) [inline store/src/test/java/org/apache/rocketmq/store/StoreStatsServiceTest.java:33]: wondering if it is necessary to start a new thread ~ will toString() fail if this service is not started ?

File: store/src/main/java/org/apache/rocketmq/store/index/IndexService.java (hunk 1 above)
Comment (lebron374) [reply store/src/main/java/org/apache/rocketmq/store/index/IndexService.java:323]: tmp may null you' ll find

File: store/src/test/java/org/apache/rocketmq/store/StoreStatsServiceTest.java (hunk 2 above)
Comment (lebron374) [reply store/src/test/java/org/apache/rocketmq/store/StoreStatsServiceTest.java:33]: if do not start a new thread，may collect nothing

File: store/src/main/java/org/apache/rocketmq/store/index/IndexService.java (hunk 1 above)
Comment (imaffe) [reply store/src/main/java/org/apache/rocketmq/store/index/IndexService.java:323]: yeah that's why we might want to separate these two conditions. If `tmp` is not null and isWriteFull, then it is likely that  another thread created this `tmp` index file, which is different from the previous `prevIndexFile`. Probably I'm wrong LOL

Comment (lebron374) [reply store/src/main/java/org/apache/rocketmq/store/index/IndexService.java:323]: o,maybe my mistake， i don't have the idea to change whole code，i just try to fix the bug without change the conditions

**Comment by lebron374** [comment]
i create new pr https://github.com/apache/rocketmq/pull/2041

**Pull request closed without merging**